### PR TITLE
186692527 update spm related non hud reports

### PR DESCRIPTION
**Pull request by ttoomey**
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description
[186692527
](https://www.pivotaltracker.com/story/show/186692527)

Also:
* Remove twenty_twenty_four_warning from views
* update exit destination codes for 2024
* adjust 1a and 1b cols/rows for off-by one from 2021 implementation (C2 => B1)

Notes:
* I adapted  Performance Metrics, CoC Performance Measurement Dashboard, and System Performance Measures by Sub-Population reports. These reports made use of the SpmClient records in 2020; those are not present in 2024 due to refactoring. While the reports run, **I have low confidence that I translated them correctly to use the new structures** (SpmClient/Return/Episode). There are no existing tests or any written specs for these reports.
* After discussion with Elliot, we decided to use the HUD data directly rather than denormalizing more values onto the Spm report records. While the HUD records are not stable, they are unlikely to change while the report is running
* The project scorecard interface seems broken, the controller expects params that are not in the form. I hacked it to get it to run.
 

## Type of change
[//]: # 'remove options that are not relevant'
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/performance_metrics/app/models/performance_metrics/report.rb [hunk 1]
```diff
@@ -607,13 +607,18 @@ def report_labels
         spm_returners = answer_clients(spm_report, '2', 'I7')
         # M2 I7 is Total Number of Persons who Exited to a Permanent Housing Destination (2 Years Prior)
         spm_leavers = answer_clients(spm_report, '2', 'B7')
+
+        # FIXME, the leavers/returners (universe_members) are SpmEnrollments in 2024; they used to be SpmClients.
+        # We probably need to to unpack enrollment.episodes to get equiv of m1a_es_sh_th_days and m2_reentry_days.
```
Comment (ttoomey) [inline drivers/performance_metrics/app/models/performance_metrics/report.rb]: I might need some advice on how to get the equivalent fields from SPM Enrollment

File: drivers/performance_metrics/app/models/performance_metrics/report.rb [hunk 2]
```diff
@@ -676,20 +682,15 @@ def report_labels
 
     private def run_spm
       # puts 'Running SPM'
-      # Looking for SPM Measure 1A E3
-      # Looking for SPM Measure 2 J7 (total returns to homelessness within 2 years)
-      questions = [
-        'Measure 1',
-        'Measure 2',
-      ]
+      questions = ['Measure 2']
```
Comment (ttoomey) [inline drivers/performance_metrics/app/models/performance_metrics/report.rb]: doesn't seem like measure 1 is used, comment seems out of date

Comment (eanders) [reply drivers/performance_metrics/app/models/performance_metrics/report.rb]: Totally possible.

File: app/views/warehouse_reports/_twenty_twenty_four_warning.haml
Comment (eanders) [inline app/views/warehouse_reports/_twenty_twenty_four_warning.haml:1]: We may need to leave this in place if we are not updating the performance_measurement report yet.

Comment (ttoomey) [reply app/views/warehouse_reports/_twenty_twenty_four_warning.haml:1]: Yeah, if we want to do this in phases we'd still need to banner on the metrics/measures reports until we fix them. Or we could try and update them all in one pass, either way is fine with me

File: drivers/performance_measurement/app/models/performance_measurement/report.rb [hunk 3]
```diff
@@ -259,6 +259,19 @@ def can_see_client_details?(user)
...
     end
 
+    private def spm_enrollments_from_answer_member(member)
+      case member
+      when HudSpmReport::Fy2023::SpmEnrollment
+        [member]
+      when HudSpmReport::Fy2023::Episode
+        member.enrollments
+      when HudSpmReport::Fy2023::Return
+        [member.return_enrollment, member.exit_enrollment]
```
Comment (ttoomey) [inline drivers/performance_measurement/app/models/performance_measurement/report.rb:269]: maybe this should just be the exit enrollment?

**Review by eanders — APPROVED** [review]
@ttoomey thank you so much for this!  It looks great.

**Comment by eanders** [comment]
Oh, looks like there are actually a few tests for this.  I'm holding off on merging until we circle back to see why those are failing.

**Comment by ttoomey** [comment]
> Oh, looks like there are actually a few tests for this. I'm holding off on merging until we circle back to see why those are failing.

I'll check it out, thanks!

**Comment by ttoomey** [comment]
> > Oh, looks like there are actually a few tests for this. I'm holding off on merging until we circle back to see why those are failing.
> 
> I'll check it out, thanks!

@eanders seems like we were didn't connect the client filters (race, age, etc). That should be corrected now in c13e736953b545783be453c61ca4592e1a808f32. However there are still test failures, it looks like we uncommented tests for `hispanic_latinaeo` filters but our sample Client.csv does not include matching clients. Not sure how that sample data was generated; should we change the tests to match the data or vice versa?

File: app/models/grda_warehouse/service_history_enrollment.rb [hunk 4]
```diff
@@ -4,6 +4,10 @@
 # License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
 ###
 
+# = GrdaWarehouse::ServiceHistoryEnrollment
+#
+# ServiceHistoryEnrollments flatten Hud Enrollments and related records to serve reporting needs. These records are
+# generated automatically. There is a 1:1 correspondence with Hud Enrollment records
```
Comment (eanders) [inline app/models/grda_warehouse/service_history_enrollment.rb:10]: This is only partially true.  There should be a ServiceHistoryEnrollment with a `record_type=entry` that corresponds to the Enrollment record, one with `record_type=exit` that corresponds to the Exit (if the client has an Exit record) and potentially one with `record_type=first` if the enrollment is the client's first homeless enrollment.  So there is a 1:1 with Enrollment for `GrdaWarehouse::ServiceHistoryEnrollment.entry`

File: drivers/hud_spm_report/app/models/hud_spm_report/adapters/service_history_enrollment_filter.rb [hunk 5]
```diff
@@ -4,6 +4,10 @@
 # License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
 ###
 
+# = HudSpmReport::Adapters::ServiceHistoryEnrollmentFilter
+#
+# Allow us to use the existing report form to filter SPM reports. Needed since the existing form filters
+# ServiceHistoryEnrollments but the SPM report operates on raw HUD data
```
Comment (eanders) [inline drivers/hud_spm_report/app/models/hud_spm_report/adapters/service_history_enrollment_filter.rb:10]: Ah!  Good find.

**Comment by eanders** [comment]
> Referenc

Thanks @ttoomey , I uncommented those since they said they _should_ be functional once we get to the FY2023 SPM.  I'll take a look and see if there's an easy fix.  Probably just need to update the source data.  If all the other tests are passing, that's excellent!

**Pull request merged**